### PR TITLE
Implicit type conversions in ISLE

### PR DESCRIPTION
**Pull request by cfallin**
An initial sketch of the ISLE compiler implementation for implicit type conversions (#3753). I can clean it up a bit but this is the general shape at least.

Two implementation bits: (i) I opted to add conversions as their own IR node type, rather than try lowering these to extractors/constructors to share that code, as the operations are just a tiny bit different (not associated with a term, for example); and (ii) there is the option to denote a converter as returning a borrow or a value -- the former is useful in patterns, the latter may be necessary if the converted value does not already exist inside the original or in the context, and we also always clone to a value when in the right-hand side to avoid holding a borrow on the mut context.

I'll work a bit further ahead by trying to use this to simplify some of the lowering rules, to make sure it actually works as we hope it will (but any such use will be merged by a later PR).

cc @fitzgen for thoughts?

### CODE REVIEW

**Comment by cfallin** [comment]
Ah, one other high-level thing: it may be useful in the future to have a mode in which we dump the sema tree, with conversions included, back to ISLE source with all the conversions made explicit. Call this "core ISLE" or a sort of MIR, I suppose. This would primarily be useful for tools that consume ISLE (e.g. for verification), so they don't need an independent implementation of the conversion-insertion logic.

Possibly it's better for such tools to just consume the sema nodes directly though -- cc @avanhatt for thoughts on this?

**Comment by avanhatt** [comment]
@cfallin thanks for the heads up! The verification engine is currently consuming sema nodes if I understand the distinction correctly (`isle::sema::{Rule, TermEnv, TypeEnv, Pattern, ...}`, code is [here](https://github.com/avanhatt/wasmtime/blob/isle-veri-notes/cranelift/isle/veri/veri_engine/src/main.rs#L15)). I think seeing an example of how this would affect an existing lowering rule would help me understand the verification implications; even if that actual change doesn't happen in this PR.

One thing I'm not quite understanding from this PR yet - in the example converter `(convert T U t_to_u)`, is the assumption that this requires that the external Rust type T implements `Into<U>`? Or is there some other way of defining the Rust implementation of the converter?

Also, can multiple converters be chained? That is, if I have `a_to_b` and `b_to_c`, can I use an `a` as a `c` without a direct converter?

**Comment by cfallin** [comment]
> @cfallin thanks for the heads up! The verification engine is currently consuming sema nodes if I understand the distinction correctly (`isle::sema::{Rule, TermEnv, TypeEnv, Pattern, ...}`, code is [here](https://github.com/avanhatt/wasmtime/blob/isle-veri-notes/cranelift/isle/veri/veri_engine/src/main.rs#L15)). I think seeing an example of how this would affect an existing lowering rule would help me understand the verification implications; even if that actual change doesn't happen in this PR.

OK, cool, I think everything should be fine then; you should be able to follow the type errors once updated past this PR with the new nodes (`sema::Pattern::Convert`, `sema::Expr::Convert`) and handle them like other external calls.

Propagation of this into the lowering rules will definitely come next! Just wanted to get the compiler implementation out for thoughts/comments first :-)

> One thing I'm not quite understanding from this PR yet - in the example converter `(convert T U t_to_u)`, is the assumption that this requires that the external Rust type T implements `Into<U>`? Or is there some other way of defining the Rust implementation of the converter?

Everything should be explicit I think (no `Into` required) -- `t_to_u` can implement arbitrary logic, like other external extractors or constructors. Importantly it has access to the context too, so it can e.g. automatically fill-in calls to `def_inst` or the like in instruction patterns.

> Also, can multiple converters be chained? That is, if I have `a_to_b` and `b_to_c`, can I use an `a` as a `c` without a direct converter?

I thought about this a bit but right now I want to see how far I can get without it. If we do need this, then we could compute the transitive closure of all conversions before consulting the conversions table during typechecking, or we could ask the user to define A->C explicitly; I think I'd slightly prefer the latter as it simplifies codegen.

**Comment by fitzgen** [comment]
> > Also, can multiple converters be chained? That is, if I have `a_to_b` and `b_to_c`, can I use an `a` as a `c` without a direct converter?
> 
> I thought about this a bit but right now I want to see how far I can get without it. If we do need this, then we could compute the transitive closure of all conversions before consulting the conversions table during typechecking, or we could ask the user to define A->C explicitly; I think I'd slightly prefer the latter as it simplifies codegen.

Yeah I would definitely prefer not to handle transitive conversions, if we can.

**Comment by fitzgen** [comment]
> I opted to add conversions as their own IR node type, rather than try lowering these to extractors/constructors to share that code, as the operations are just a tiny bit different (not associated with a term, for example)

It seems that they *are* associated with a term: the term specified in the conversion declaration.

-------------------------

Also, is this separating conversions for extractors vs constructors, like we talked about in the issue?

File: cranelift/isle/isle/src/ast.rs
Comment (fitzgen) [inline cranelift/isle/isle/src/ast.rs:439]: Can this be inferred from whether the term used for converting is infallible?

File: cranelift/isle/isle/src/ast.rs
Comment (fitzgen) [inline cranelift/isle/isle/src/ast.rs:442]: It seems strange to me that we need/have this for conversions but not for regular terms. Every place where we want to insert these implicit conversions works right now and doesn't have this `borrow` flag, so I feel like it shouldn't be necessary to add. I must be missing something here.

File: cranelift/isle/isle/src/ast.rs
Comment (fitzgen) [inline cranelift/isle/isle/src/ast.rs:430]: Is this assuming that the term used for converting is always an external Rust function? That seems like an unnecessary restriction to me.

File: cranelift/isle/isle/src/ast.rs
Comment (fitzgen) [inline cranelift/isle/isle/src/ast.rs:428]: This also doesn't seem to differentiate between conversions in an extraction vs construction context, which we talked about in the issue.

**Comment by cfallin** [comment]
> > I opted to add conversions as their own IR node type, rather than try lowering these to extractors/constructors to share that code, as the operations are just a tiny bit different (not associated with a term, for example)
> 
> It seems that they _are_ associated with a term: the term specified in the conversion declaration.

I'm not sure I follow, do you mean the Rust function name? E.g. in specifying `(convert T U t_to_u)` there are two types, and a Rust function, but no term involved.

> Also, is this separating conversions for extractors vs constructors, like we talked about in the issue?

I actually started down that path but found there was a lot of unnecessary duplication: it's simpler just to have one mechanism for going from T to U I think. Or said another way, once we have `t_to_u` and the system knows about it, there's no difference in behavior whether we use that to take a `T` and apply a pattern for a `U` on the extractor side, or evaluate an expression for a `T` and convert it to a final `U` result on the constructor side. But possibly I'm missing some reason why we'd want to disallow a particular conversion on one side or the other; do you have an example in mind?

File: cranelift/isle/isle/src/sema.rs
Comment (fitzgen) [inline cranelift/isle/isle/src/sema.rs:720]: This can avoid a double lookup by doing something like

```rust
let old_entry = tyenv.converters.insert(...);
if let Some(old_entry) = old_entry {
    tyenv.report_error(...);
    continue;
}
```

File: cranelift/isle/isle_examples/conversions.isle
Comment (fitzgen) [inline cranelift/isle/isle_examples/conversions.isle:1]: This is unrelated to these particular changes, but long overdue nonetheless: can we write a `#[test]` that asserts that every file in `isle_examples` compiles okay? (At least through ISLE, testing the emitted code with `rustc` will be a bit harder/more tedious since it requires glue code).

And for the ones that are supposed to fail (I forget if we even have any of those) it would be nice if we had a comment at the top with a substring of the expected error message. These could be in a separate directory and also be handled in separate PRs.

But we really need *some* kind of testing story for this stuff, especially as we tweak the language and compiler.

**Review by fitzgen — COMMENTED** [review]
I like the direction for sure, but I think that the conversions should be regular terms, instead of always being external functions. This has a number of benefits:

* It would allow us to write converters inline in ISLE, rather than always in external Rust code. This would be useful when converting between types that are defined in ISLE itself, rather than externally.
* It would still allow us to support external conversions via external constructors/extractors.
* This would allow us to reuse the existing infallibility support we have for external constructors/extractors.
* We could potentially use this to differentiate between when we are converting inside an extraction vs construction context by using the terms constructor or extractor (we would have to check not just that conversions exist but that the associated term has an extractor/constructor in type checking; also the conversion lookup would need to be a tiny bit more complicated to deal with how extractors go "backwards")
* It allows us to simplify the IR since we can insert regular constructor/extractor term calls for conversions that we already support instead of defining a new `Convert` variant like this PR currently does. Similarly, it means that code generation won't need any changes at all, the implicit conversions are strictly a front-end concern.

Thoughts?

**Comment by fitzgen** [comment]
> But possibly I'm missing some reason why we'd want to disallow a particular conversion on one side or the other; do you have an example in mind?

In an extractor, it maybe makes sense to implicitly convert from a `Value` to an `Inst` via the `def_inst` extractor. But I don't think I would want to do that same implicit conversion in constructors. Hard to put into words why I'm reacting that way. It just feels wrong/unnecessary/I wouldn't use it that way, and so I'd rather not have any kind of potential footgun there.

FWIW, 99% of the sites where I personally wish we had implicit conversions are constructors.

File: cranelift/isle/isle_examples/conversions.isle
Comment (cfallin) [reply cranelift/isle/isle_examples/conversions.isle:1]: Yes, definitely agreed, this is probably as good a time as any to start!

File: cranelift/isle/isle/src/ast.rs
Comment (cfallin) [reply cranelift/isle/isle/src/ast.rs:439]: Yes, if we go to a term-based design instead, for sure.

**Comment by fitzgen** [comment]
To clarify a little bit on the semantics/lookup I'm imagining for implicit conversions:

```lisp
;; `convert_T_U`'s constructor is used when implicitly converting
;; T to U in a constructor context.
;;
;; `convert_T_U`'s extractor is used when implicitly converting U
;; to T in an extractor context.
(convert T U convert_T_U)

(decl convert_T_U (T) U)

;; convert T to U in constructors.
(rule (convert_T_U t) ...)

;; convert U to T in extractors.
(extractor (convert_T_U t) ...)
```

Does that make sense?

File: cranelift/isle/isle/src/ast.rs
Comment (cfallin) [reply cranelift/isle/isle/src/ast.rs:442]: So this became necessary because conversion functions can be used on both the LHS and RHS, and on the LHS we want to deal in borrows as much as possible (as extractors work today) while on the RHS we're usually building entirely new structs. If we were to lower to terms instead then it wouldn't be necessary, as the extractor/constructor codegen already does basically the right thing here.

**Comment by cfallin** [comment]
> > But possibly I'm missing some reason why we'd want to disallow a particular conversion on one side or the other; do you have an example in mind?
> 
> In an extractor, it maybe makes sense to implicitly convert from a `Value` to an `Inst` via the `def_inst` extractor. But I don't think I would want to do that same implicit conversion in constructors. Hard to put into words why I'm reacting that way. It just feels wrong/unnecessary/I wouldn't use it that way, and so I'd rather not have any kind of potential footgun there.
> 
> FWIW, 99% of the sites where I personally wish we had implicit conversions are constructors.

This is a really interesting difference -- I found that my intuition was actually pushing the other way! Basically it feels like a feature of the low-level dataflow in the language to me; typechecking can insert little converters to make things work out where there are gaps between the type, and this is below the level of concern of extractors/constructors. I think most the other comments above flow from that difference as well.

In general as well, I want to try to avoid distinctions between extractors and constructors where possible; if we instead see the forms on LHS and RHS as living in the same value-space, with similar semantics, it opens the door later to optimizations like chaining multiple rewrite steps into one. In that sense, it feels like an odd semantic gap for a conversion from A to B to be possible on the LHS, but not on the RHS ("one rule's LHS is another rule's RHS" in some sense, once we start to chain steps).

In the case of `Value` to `Inst`, you do raise a good point, the meaning of that is sort of odd if it's invoke on the RHS. Perhaps the right way to resolve this though is to say that fallible conversions can only appear in the LHS. `def_inst` is fallible because we can't always peek back at the defining instruction for a value; so the `Value` to `Inst` conversion is fallible, and wouldn't make sense on the RHS. Does that seem reasonable?

**Comment by cfallin** [comment]
Ah, but re: explicit calls to external Rust functions vs. just inserting terms, I do agree completely -- delegation to the outside should be an orthogonal thing, and lowering to "core ISLE" that really is just etors/ctors makes sense as a separate desugaring step. I'll start to rework things in that direction.

File: cranelift/isle/isle/src/ast.rs
Comment (cfallin) [reply cranelift/isle/isle/src/ast.rs:428]: I don't think this is needed here: even if we do move to a terms-as-converters rather than explicit-converter-funcs design, we can look up the term and find an attached extractor or converter, so we don't need to record directionality here, I think.

File: cranelift/isle/isle/src/ast.rs
Comment (cfallin) [reply cranelift/isle/isle/src/ast.rs:430]: (commented below -- yes, I'll change this)

File: cranelift/isle/isle/src/ast.rs
Comment (cfallin) [reply cranelift/isle/isle/src/ast.rs]: Removed now, with term-based design.

File: cranelift/isle/isle/src/sema.rs
Comment (cfallin) [reply cranelift/isle/isle/src/sema.rs]: Thanks! Used the entry API actually for a similarly cleaner single-lookup.

File: cranelift/isle/isle_examples/conversions.isle
Comment (cfallin) [reply cranelift/isle/isle_examples/conversions.isle]: Done! We now have pass-tests, fail-tests, and link-tests; the last one checks that the generated code builds and links against the example driver/context trait impl.

**Comment by cfallin** [comment]
OK, so I've reworked this as per the above and after talking with @fitzgen a bit more today -- it's much cleaner now, I think, in that conversions are immediately lowered to ordinary ctor/etor invocations. I added some test infrastructure as well. (The fail-tests don't assert on the actual failing error message yet; maybe we could do something with filetest; but I'll leave that to a future PR.)

I'm still planning to try using this for a bit before merging the compiler bits; I'll switch this out of "Draft" once I've convinced myself it's a reasonable and usable feature.

**Comment by cfallin** [comment]
OK, so I think this is ready for review! cc @fitzgen 

I took this for a spin in [this branch (x64/lower.isle)](https://github.com/cfallin/wasmtime/blob/isle-implicit-conversions-rules/cranelift/codegen/src/isa/x64/lower.isle), which builds on the ISLE compiler changes here. I'm quite excited by how clean the rules turn out to be -- for example, the basic lowering for integer add is literally

```lisp
;; Add two registers.
(rule (lower (has_type (fits_in_64 ty)
                       (iadd x y)))
      (add ty x y))
```

where the `(put_value_in_reg ...)` and `(value_reg ...)` are inserted implicitly. All of the conversions between Gpr/Xmm newtypes, the various argument forms (Reg, RegMem, RegMemImm, likewise for Xmm/Gpr), etc. all became automatic as well.

This PR has just the changes to the DSL compiler, and documentation changes, as well as the new ISLE testing infrastructure. I'll do a followup PR with the branch linked above containing a refactor to actually define and use some converters.

**Review by fitzgen — APPROVED** [review]
LGTM! Really happy with the way this turned out.

r=me with nitpicks below

Might also be worth running the ISLE fuzz target for an hour or so just to shake out any lurking bugs (since we don't run this fuzz target in oss-fuzz)

**Comment by avanhatt** [comment]
FWIW, I tried merging this with our verification branch to try it on the iadd example (full test case [here](https://github.com/avanhatt/wasmtime/blob/veri-isle-conversions/cranelift/isle/veri/veri_engine/src/main.rs#L145)), and I'm seeing this newly added error from `isle::sema` in using a constructor. Am I messing up the syntax here? Also, should I be seeing the name of the term that's unknown?

```rust
called `Result::unwrap()` on an `Err` value: TypeError { msg: "Unknown term for converter", src: Source { name: "fuzz-input.isle", source: "<redacted>", span: SourceSpan { offset: SourceOffset(362), length: SourceOffset(1) } }, cranelift/isle/veri/veri_engine/src/main.rs:26:58
```

Example snippet (full snippet linked above):
```rust
        (type Inst (primitive Inst))
        (type Type (primitive Type))
        (type Value (primitive Value))
    
        (type Reg (primitive Reg))
        (type ValueRegs (primitive ValueRegs))

        ;; NEW: IMPLICIT CONVERTERS
        (convert Reg ValueRegs value_reg)
        (convert Value Reg put_in_reg)
        
        ;; ...
        (rule (lower (has_type (fits_in_64 ty) (iadd x y)))
            (add ty x y))
```

File: cranelift/isle/isle/src/sema.rs
Comment (cfallin) [reply cranelift/isle/isle/src/sema.rs]: Done!

File: cranelift/isle/isle/isle_examples/fail/bad_converters.isle
Comment (cfallin) [reply cranelift/isle/isle/isle_examples/fail/bad_converters.isle:10]: Added!

File: cranelift/isle/isle/build.rs
Comment (cfallin) [reply cranelift/isle/isle/build.rs:33]: Yup, I'm glad we have them now!

File: cranelift/isle/isle/Cargo.toml
Comment (cfallin) [reply cranelift/isle/isle/Cargo.toml]: Fixed!

File: cranelift/docs/isle-integration.md
Comment (cfallin) [reply cranelift/docs/isle-integration.md]: Good point, just removed this parenthetical for now; we can work out the details and document appropriately if/when we need to do this.

**Comment by cfallin** [comment]
@avanhatt that's odd; that's a fairly basic error indicating that the lookup of the term name itself (`value_reg` or `put_in_reg` in your snippet) failed. The syntax looks right, and this feature is definitely working correctly otherwise in tests and in my exploratory use of it, so I might check if the merge introduced some issue?

**Comment by avanhatt** [comment]
@cfallin ah, it was a silly mistake, I misinterpreted the `convert` statement as _replacing_ the `decl`/`extern ...` statements for the term rather than being in addition to them. Looks like it works now adding those back! 

I might add to the section of the docs that the converter must already be defined? This part is what I misunderstood:
```
Now we can define a conversion such as

 ```lisp
     (convert Reg ValueRegs value_reg)
```

**Comment by fitzgen** [comment]
Either way, that should be a nicely reported error, not an `unwrap` panic.

**Comment by fitzgen** [comment]
Oh I see it is your code that is doing the `unwrap` not the ISLE compiler itself, nvm!

**Comment by fitzgen** [comment]
@avanhatt FWIW, you should be able to get better error messages with source context snippets and stuff (like how `rustc` provides) if you crib from this code a bit:

* https://github.com/bytecodealliance/wasmtime/blob/main/cranelift/codegen/build.rs#L325
* https://github.com/bytecodealliance/wasmtime/blob/main/cranelift/codegen/build.rs#L417-L444

(It is annoying that it requires this song and dance, and we should improve that at some point, but that's what is needed for the moment.)

**Comment by avanhatt** [comment]
Great, thanks for the edits + source debugging tip! And sorry for the `unwrap` confusion. 

One other small observation about implicit conversions: one thing you might lose when these are done implicitly is seeing why a special case rewrite is beneficial at a first glance. 

For example, in the add-to-sub rule, it's not as obvious anymore that the RHS avoids an extra register use, because visually `x` and `y` look the same:

```
(rule (lower (has_type (fits_in_64 ty) (iadd x (imm12_from_negated_value y))))
      (sub_imm ty x y))
```

Definitely less important than the convenience this adds, especially with motivating comments, but thought I'd point it out.

**Comment by fitzgen** [comment]
> For example, in the add-to-sub rule, it's not as obvious anymore that the RHS avoids an extra register use, because visually `x` and `y` look the same:

Yeah there is certainly a balance to be struck here. For example, with x64's `RegMemImm`, it may make sense to only have conversions for `Reg` to `RegMemImm` and not from memory operands and immediates. This would let us highlight the interesting operands while cutting out noise for the uninteresting common case of registers.

**Pull request merged**
